apps/extension/src/routes/page/onboarding/set-grpc-endpoint.tsx

```diff
@@ -3,11 +3,26 @@ import { FadeTransition } from '@repo/ui/components/ui/fade-transition';
 import { usePageNav } from '../../../utils/navigate';
 import { PagePath } from '../paths';
 import { GrpcEndpointForm } from '../../../shared/components/grpc-endpoint-form';
+import { localExtStorage } from '../../../storage/local';
+import { AppParameters } from '@penumbra-zone/protobuf/penumbra/core/app/v1/app_pb';
 
 export const SetGrpcEndpoint = () => {
   const navigate = usePageNav();
 
-  const onSuccess = (): void => {
+  // For beta testing: set the wallet block height to zero for non-mainnet chain IDs.
+  // This logic only runs after the user selects their rpc endpoint.
+  const onSuccess = async (): Promise<void> => {
```

Comment:
issue: unfortunately, I still think this is running after services has initialized. Sadly it looks like the `useGrpcEndpointForm` is incredibly complicated. It's an example of where useEffect() adds complexity... See:

- [setGrpcEndpoint](https://github.com/prax-wallet/prax/blob/3ee1ff6954677d0862623c930f3ee985069496a2/apps/extension/src/shared/components/grpc-endpoint-form/use-grpc-endpoint-form.ts#L132-L138) runs before onSuccess
- which means local storage [being set first](https://github.com/prax-wallet/prax/blob/3ee1ff6954677d0862623c930f3ee985069496a2/apps/extension/src/state/network.ts#L25)
- which means [startWalletServices()](https://github.com/prax-wallet/prax/blob/3ee1ff6954677d0862623c930f3ee985069496a2/apps/extension/src/wallet-services.ts#L16) proceeds through that line and starts services with the old walletCreationBlockHeight

Reply:
very tricky – only thing that comes to mind as a potential solution is refactoring the `useEffect` to delay service initialization using a flag where `startWalletServices` only runs after `walletCreationBlockHeight` is set.

Reply:
or maybe move the logic directly into [onSubmit](https://github.com/prax-wallet/prax/blob/3ee1ff6954677d0862623c930f3ee985069496a2/apps/extension/src/shared/components/grpc-endpoint-form/use-grpc-endpoint-form.ts#L112)?

Reply:
like why can't we reorder the operations inside the [onSubmit](https://github.com/prax-wallet/prax/blob/3ee1ff6954677d0862623c930f3ee985069496a2/apps/extension/src/shared/components/grpc-endpoint-form/use-grpc-endpoint-form.ts#L112) async function?

```ts
        const storedParams = await localExtStorage.get('params');
        if (storedParams) {
          const parsedParams = JSON.parse(storedParams) as AppParameters;
          if (!parsedParams.chainId.includes('penumbra-1')) {
            await localExtStorage.set('walletCreationBlockHeight', 0);
          }
        }
    

       await onSuccess();
```
    
which ensures a proper sequencing?

Reply:
there must be a way to explicitly block on wallet service init until this code block completes.

Reply:
or maybe move `await onSuccess()` before `void chrome.runtime.sendMessage(ServicesMessage.ClearCache)`? does that functionally break anything?

Reply:
or we can block wallet service init directly inside [onboardGrpcEndpoint](https://github.com/prax-wallet/prax/blob/3ee1ff6954677d0862623c930f3ee985069496a2/apps/extension/src/wallet-services.ts#L16)?

Reply:
settled on https://github.com/prax-wallet/prax/pull/216/commits/fcd8e41e16de32ac4916e6a2ece2e48916543ea4 for simplicity and readability.